### PR TITLE
Add check for doc alias attribute at crate level

### LINKED ISSUE #76298: `#![doc(alias = "...")]` is ignored as a crate level attribute
Either the alias should show up the search results like aliases on other things or it should be disallowed.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #76298, #64734, #69365.

r? @ollie27

### CODE REVIEW

File: compiler/rustc_passes/src/check_attr.rs [hunk 1]
```diff
@@ -518,8 +542,12 @@ fn is_c_like_enum(item: &Item<'_>) -> bool {
...
 
 fn check_mod_attrs(tcx: TyCtxt<'_>, module_def_id: LocalDefId) {
-    tcx.hir()
-        .visit_item_likes_in_module(module_def_id, &mut CheckAttrVisitor { tcx }.as_deep_visitor());
+    let mut visitor = CheckAttrVisitor { tcx };
+    tcx.hir().visit_item_likes_in_module(module_def_id, &mut visitor.as_deep_visitor());
+    let hir_id = tcx.hir().local_def_id_to_hir_id(module_def_id);
+    for attr in tcx.hir().krate_attrs() {
+        visitor.check_doc_alias(attr, hir_id, Target::Mod, true);
+    }
```
Comment (ollie27) [inline compiler/rustc_passes/src/check_attr.rs:550]: `check_mod_attrs` is called for each module in a crate so isn't this going to emit the error for every module in the crate?

Comment (GuillaumeGomez) [reply compiler/rustc_passes/src/check_attr.rs:550]: Not the crate module apparently. I tested on this code:

```rust
#![feature(doc_alias)]
#![doc(alias = "test")]
```

Never saw an inner attribute. That's how I discovered that the krate module was passed down.

Comment (ollie27) [reply compiler/rustc_passes/src/check_attr.rs:550]: So `check_mod_attrs` is called for every module in the crate including the crate root module but the problem is that it only checks items inside the module and not the module itself. This looks like a preexisting issue in rustc and I'm not sure how best to fix it. What is here though will check the crate attributes multiple times which is definitely not right.

I'm not actually convinced that we should disallow `#![doc(alias)]` for the crate root module as it is allowed on all other modules.

Comment (GuillaumeGomez) [reply compiler/rustc_passes/src/check_attr.rs:550]: I think it's intended this way.

> What is here though will check the crate attributes multiple times which is definitely not right.

What do you mean? As far as I can tell, it checks the crates' attributes once and that's it. Did I miss something?

> I'm not actually convinced that we should disallow `#![doc(alias)]` for the crate root module as it is allowed on all other modules.

In which case would we want to have a doc alias on the crate module itself?

Comment (ollie27) [reply compiler/rustc_passes/src/check_attr.rs:550]: > I think it's intended this way.

rustc not checking crate level attributes is very much a bug causing #64734 and #69365 for example. I would like to see a proper fix for this.

> What do you mean? As far as I can tell, it checks the crates' attributes once and that's it. Did I miss something?

You can see from where `check_mod_attrs` is called that it is called once per module in the crate: https://github.com/rust-lang/rust/blob/da897dfb6daa268a965d2d73620316ba56020a19/compiler/rustc_interface/src/passes.rs#L812-L818 and https://github.com/rust-lang/rust/blob/445f34bb144c5b163163d8243e9845c644d3662d/src/librustdoc/core.rs#L514-L517

That means errors are emitted multiple times. This can be seen be compiling a crate with this PR applied like:

```rust
#![feature(doc_alias)]
#![doc(alias = "hello")]
mod foo {}
```

When `-Zdeduplicate-diagnostics=no` is passed to rustc or rustdoc the output is:

```
error: `#![doc(alias = "...")]` isn't allowed as a crate level attribute
 --> foo.rs:2:8
  |
2 | #![doc(alias = "hello")]
  |        ^^^^^^^^^^^^^^^

error: `#![doc(alias = "...")]` isn't allowed as a crate level attribute
 --> foo.rs:2:8
  |
2 | #![doc(alias = "hello")]
  |        ^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors
```

Adding more modules causes the error to be repeated even more.

> In which case would we want to have a doc alias on the crate module itself?

I have no idea, but I also don't know why you would want doc aliases on any module. I'm fine with disallowing it because we can always allow it in the future if someone has a use case.

Comment (GuillaumeGomez) [reply compiler/rustc_passes/src/check_attr.rs:550]: Actually, we need them for modules because of primitives. :)

**Comment by GuillaumeGomez** [comment]
@ollie27 More information: it's actually pretty complicated to pass down the crate level because we try to run its components in parallel or want to filter the items we iterate over. I'm not sure there is a global fix for this, or at least don't see any obvious one...

**Comment by GuillaumeGomez** [comment]
ping @ollie27

File: compiler/rustc_passes/src/check_attr.rs [hunk 2]
```diff
@@ -270,6 +270,16 @@ impl CheckAttrVisitor<'tcx> {
...
                         }
+                        if attr.style == AttrStyle::Inner && CRATE_HIR_ID == hir_id {
+                            self.tcx
+                                .sess
+                                .struct_span_err(
+                                    meta.span(),
+                                    "`#![doc(alias = \"...\")]` isn't allowed as a crate \
+                                     level attribute",
+                                )
+                                .emit();
```
Comment (ollie27) [inline compiler/rustc_passes/src/check_attr.rs:281]:
```suggestion
                                .emit();
                            return false;
```

If we emit an error then we should return `false` like the other attribute checks do.

File: compiler/rustc_passes/src/check_attr.rs [hunk 3]
```diff
@@ -270,6 +270,16 @@ impl CheckAttrVisitor<'tcx> {
                                 )
                                 .emit();
                         }
+                        if attr.style == AttrStyle::Inner && CRATE_HIR_ID == hir_id {
```
Comment (ollie27) [inline compiler/rustc_passes/src/check_attr.rs:273]:
```suggestion
                        if CRATE_HIR_ID == hir_id {
```

The attribute style check isn't needed.

File: compiler/rustc_passes/src/check_attr.rs [hunk 4]
```diff
@@ -530,6 +540,11 @@ fn is_c_like_enum(item: &Item<'_>) -> bool {
 fn check_mod_attrs(tcx: TyCtxt<'_>, module_def_id: LocalDefId) {
     tcx.hir()
         .visit_item_likes_in_module(module_def_id, &mut CheckAttrVisitor { tcx }.as_deep_visitor());
+    if tcx.hir().local_def_id(CRATE_HIR_ID) == module_def_id {
+        for attr in tcx.hir().krate_attrs() {
+            CheckAttrVisitor { tcx }.check_doc_alias(attr, CRATE_HIR_ID, Target::Mod);
+        }
```
Comment (ollie27) [inline compiler/rustc_passes/src/check_attr.rs:546]: We could call [`check_attributes`](https://github.com/rust-lang/rust/blob/07ece44a42b6cd33e59d58b5325f637f398464d2/compiler/rustc_passes/src/check_attr.rs#L52) here which should fix #64734 and #69365 while we're at it.

If you want to just apply this to the `#[doc]` attribute for now though this will need a `if self.tcx.sess.check_name(attr, sym::doc)` check.

File: compiler/rustc_passes/src/check_attr.rs [hunk 5]
```diff
@@ -530,6 +540,11 @@ fn is_c_like_enum(item: &Item<'_>) -> bool {
 fn check_mod_attrs(tcx: TyCtxt<'_>, module_def_id: LocalDefId) {
     tcx.hir()
         .visit_item_likes_in_module(module_def_id, &mut CheckAttrVisitor { tcx }.as_deep_visitor());
+    if tcx.hir().local_def_id(CRATE_HIR_ID) == module_def_id {
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply compiler/rustc_passes/src/check_attr.rs:543]: :+1:

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
ping @ollie27

File: src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.rs [hunk 6]
```diff
@@ -7,9 +7,6 @@
 
 // ignore-tidy-linelength
 
-// Crate-level is accepted, though it is almost certainly unused?
-#![inline]
```
Comment (ollie27) [inline src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.rs:11]: Did this attribute need to be removed? If this now produces as error then it should stay so the UI test can make sure it stays an error.

File: src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs.rs [hunk 7]
```diff
@@ -52,20 +55,8 @@
...
-#![main] //~ WARN unused attribute
-#![start] //~ WARN unused attribute
-// see issue-43106-gating-of-test.rs for crate-level; but non crate-level is below at "4200"
-// see issue-43106-gating-of-bench.rs for crate-level; but non crate-level is below at "4100"
-#![repr()]
-//~^ WARN unused attribute
-#![path = "3800"] //~ WARN unused attribute
-#![automatically_derived] //~ WARN unused attribute
-#![no_mangle]
-#![no_link] //~ WARN unused attribute
```
Comment (ollie27) [inline src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs.rs:67]: Why were these removed? If they're now errors then they should be moved to `issue-43106-gating-of-builtin-attrs-error.rs` or another dedicated test to make sure they stay errors.

**Comment by ollie27** [comment]
Apart from a few questions about the test changes this looks good to me.

I'm going to hand this over this over to the compiler team though because this is not really a rustdoc change anymore and is introducing new errors. Also to make sure we haven't missed an obvious better way to check crate level attributes.

r? @matthewjasper

File: src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs.rs:67]: It's a `check-pass` kind of test. Creating another one test including all the newly failing ones.

**Comment by GuillaumeGomez** [comment]
I put back the checks into another file which is expected to fail.

File: src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.stderr [hunk 8]
```diff
@@ -53,84 +53,151 @@ LL | |
...
+LL | |
+LL | |
+LL | |
+...  |
+LL | | #![no_mangle]
+   | | ^^^^^^^^^^^^^
+...  |
+LL | |
+LL | | fn main() {}
+   | |____________- not a function or static
```
Comment (ollie27) [inline src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.stderr:69]: Highlighting the entire crate span is a bit confusing here. I've not tested but maybe passing `DUMMY_SP` to `check_attributes` instead of `tcx.hir().span(CRATE_HIR_ID)` might work to clean this up.

File: src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.rs [hunk 9]
```diff
@@ -7,9 +7,28 @@
...
+//~^^^ NOTE not an `extern crate` item
+//~^^^^ NOTE not a function or static
+//~^^^^^ NOTE not a function or closure
+
+#![macro_export]
+#![main]
+#![start]
+#![repr()]
+#![path = "3800"]
+#![automatically_derived]
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.rs:22]: I think it's a bug from the json error interpreter: if you look at the stderr file, you can see them, however they all point to the `#[deny()]` attribute.

[the remark this answers is not in the document]
Comment (ollie27) [reply src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.rs:22]: I don't see them in the the stderr file. I'm talking specifically about these:

```rust
#![macro_export]
#![main]
#![start]
#![repr()]
#![path = "3800"]
#![automatically_derived]
```

The notes attached to the `#[deny()]` attribute are just the notes that are applied to the crate span which I was refering to in https://github.com/rust-lang/rust/pull/76329/files#r499083161.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/ui/feature-gate/issue-43106-gating-of-builtin-attrs-error.rs]: I added a specific check for the crate-level attributes that shouldn't be there.

**Comment by GuillaumeGomez** [comment]
Updated!
